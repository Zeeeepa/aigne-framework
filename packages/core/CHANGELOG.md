## [1.3.0](https://github.com/AIGNE-io/aigne-framework/compare/core-v1.2.0...core-v1.3.0) (2025-03-24)


### Features

* add Agent FunctionAgent AIAgent MCPAgent and ExecutionEngine ([4d2a5a1](https://github.com/AIGNE-io/aigne-framework/commit/4d2a5a1b3366b8f935f50a0937c2da6e49073348))
* add OrchestratorAgent in agent library ([25a5e9e](https://github.com/AIGNE-io/aigne-framework/commit/25a5e9e6c60d747c8bf484ac884b31dc02c14757))
* add sequential and parallel helper function ([a295697](https://github.com/AIGNE-io/aigne-framework/commit/a295697b5694754e02954fc5c7f382a3b219a3ab))
* add support for MCP resources ([1ded2fb](https://github.com/AIGNE-io/aigne-framework/commit/1ded2fbf222fa8984e75df0852ff384524f73b04))
* **core:** add ChatModelClaude to use models of anthropic ([#30](https://github.com/AIGNE-io/aigne-framework/issues/30)) ([0a62a64](https://github.com/AIGNE-io/aigne-framework/commit/0a62a6499e3da723a4646e67952051708ce7de6a))
* **core:** add support for subscribing topics for agent memory ([#28](https://github.com/AIGNE-io/aigne-framework/issues/28)) ([eeecc67](https://github.com/AIGNE-io/aigne-framework/commit/eeecc67049a60ebcc4cdba0fbcd987b3d81f4af6))
* **prompt-builder:** support chat history in PromptBuilder ([6ca05f2](https://github.com/AIGNE-io/aigne-framework/commit/6ca05f28eddb683a4f1e228865f8bbf8a8e190f1))
* support run puppeteer example chat loop in terminal ([85ce7f8](https://github.com/AIGNE-io/aigne-framework/commit/85ce7f8de8b443c86e50815dd7bcab99f869c4ce))
* use PromptBuilder instead of string instructions ([e4cb2cb](https://github.com/AIGNE-io/aigne-framework/commit/e4cb2cb6baf4f9bcef390567a4a174e9246c29a3))


### Bug Fixes

* **AIAgent:** should pass both arguments (model generated) and input (user input) to tool ([c49d64e](https://github.com/AIGNE-io/aigne-framework/commit/c49d64ee35f7efd83b0f82f43205bb1c40f999e8))
* **core:** enforce stricter input/output type checks ([#26](https://github.com/AIGNE-io/aigne-framework/issues/26)) ([ef8cf53](https://github.com/AIGNE-io/aigne-framework/commit/ef8cf53586aff08a809909c56ab2a20f215fa129))
* **MCP:** catch list resource error treat as empty list ([1885fab](https://github.com/AIGNE-io/aigne-framework/commit/1885fab3585e0dd1467b127e5b47cd0b98282bab))
* rename @aigne/core-next to @aigne/core ([3a81009](https://github.com/AIGNE-io/aigne-framework/commit/3a8100962c81813217b687ae28e8de604419c622))
* use text resource from MCP correctly ([8b9eba8](https://github.com/AIGNE-io/aigne-framework/commit/8b9eba83352ec096a2a5d4f410d4c4bde7420bce))

## [1.15.0](https://github.com/AIGNE-io/aigne-framework/compare/core-v1.14.0...core-v1.15.0) (2025-05-15)


### Features

* optimize the stability of the model and ci ([#119](https://github.com/AIGNE-io/aigne-framework/issues/119)) ([de93887](https://github.com/AIGNE-io/aigne-framework/commit/de938879452a8be82a198dda0eda1eb9fcbb0474))


### Bug Fixes

* **core:** response.headers.toJSON is not a function ([#121](https://github.com/AIGNE-io/aigne-framework/issues/121)) ([4609ba6](https://github.com/AIGNE-io/aigne-framework/commit/4609ba645e6b8fe8d76ecd475cd2d7817483a4bd))

## [1.14.0](https://github.com/AIGNE-io/aigne-framework/compare/core-v1.13.0...core-v1.14.0) (2025-05-12)


### Features

* **docs:** use typedoc build and publish docs to gh-pages ([#100](https://github.com/AIGNE-io/aigne-framework/issues/100)) ([b9074c0](https://github.com/AIGNE-io/aigne-framework/commit/b9074c0148ea343ada92b5919a52b47537a1ad48))
* **memory:** allow agents to act as retrievers and recorders in memory ([#65](https://github.com/AIGNE-io/aigne-framework/issues/65)) ([2bafcbb](https://github.com/AIGNE-io/aigne-framework/commit/2bafcbb66a94fcf55dad8c21ede483eaf075c11d))
* optimize the stability of ci and example ([#113](https://github.com/AIGNE-io/aigne-framework/issues/113)) ([d16ed6c](https://github.com/AIGNE-io/aigne-framework/commit/d16ed6cb60faea19fb4f1c12e1f83d69563b153f))


### Bug Fixes

* **core:** default catch tool's error and continue processing ([#115](https://github.com/AIGNE-io/aigne-framework/issues/115)) ([983b0de](https://github.com/AIGNE-io/aigne-framework/commit/983b0de491afb3f0904e145cb491d432b62f9312))
* **core:** handle response for UserAgent automatically in pub/sub mode ([#116](https://github.com/AIGNE-io/aigne-framework/issues/116)) ([b659714](https://github.com/AIGNE-io/aigne-framework/commit/b659714f2398ea042f21cb22eccc1014f181cd46))

## [1.13.0](https://github.com/AIGNE-io/aigne-framework/compare/core-v1.12.0...core-v1.13.0) (2025-04-30)


### Features

* **core:** add BedrockChatModel support ([#101](https://github.com/AIGNE-io/aigne-framework/issues/101)) ([a0b98f0](https://github.com/AIGNE-io/aigne-framework/commit/a0b98f01bd78a135232226548848fa35a64982d1))


### Bug Fixes

* **core:** deduplicate tools for chat model ([#103](https://github.com/AIGNE-io/aigne-framework/issues/103)) ([570be6d](https://github.com/AIGNE-io/aigne-framework/commit/570be6d8620ab5b9a0149f835ecd4641009a8654))
* export server/client api types ([93e5341](https://github.com/AIGNE-io/aigne-framework/commit/93e5341dde7a6851f08a3d4e2f6c1a1db91765e9))

## [1.12.0](https://github.com/AIGNE-io/aigne-framework/compare/core-v1.11.0...core-v1.12.0) (2025-04-27)


### Features

* add AIGNEServer/AIGNEClient api to serve agents via network ([#96](https://github.com/AIGNE-io/aigne-framework/issues/96)) ([1f2dfa3](https://github.com/AIGNE-io/aigne-framework/commit/1f2dfa3a6a2568373063cea3c874b573d0a248d3))
* **core:** support streamable http transport for mcp agents ([#92](https://github.com/AIGNE-io/aigne-framework/issues/92)) ([37da490](https://github.com/AIGNE-io/aigne-framework/commit/37da490538298d882ec328e4b3304395a6cd8cf7))
* support TeamAgent and finalize API naming ([#91](https://github.com/AIGNE-io/aigne-framework/issues/91)) ([033d1b6](https://github.com/AIGNE-io/aigne-framework/commit/033d1b6a7dc5460807476abb35a413ba89a2a664))


### Bug Fixes

* **core:** prioritize self model before falling back to context ([#97](https://github.com/AIGNE-io/aigne-framework/issues/97)) ([2a3d067](https://github.com/AIGNE-io/aigne-framework/commit/2a3d067442200657d8ef3b5314930cc14302f6bf))
* upgrade to streamable api for serve mcp command ([#98](https://github.com/AIGNE-io/aigne-framework/issues/98)) ([ae32bda](https://github.com/AIGNE-io/aigne-framework/commit/ae32bda20e57c2a2eb8b49fad034b0b2a5ebb15e))

## [1.11.0](https://github.com/AIGNE-io/aigne-framework/compare/core-v1.10.0...core-v1.11.0) (2025-04-23)


### Features

* **core:** enhance ClaudeChatModel to support streaming responses ([#85](https://github.com/AIGNE-io/aigne-framework/issues/85)) ([5433240](https://github.com/AIGNE-io/aigne-framework/commit/5433240e7b663ec9e9f4a79dffa05038088d54fc))
* support set memory in agent yaml ([#90](https://github.com/AIGNE-io/aigne-framework/issues/90)) ([215118f](https://github.com/AIGNE-io/aigne-framework/commit/215118f1dc55f02322d59a3f18395a459198e031))


### Bug Fixes

* **core:** router model should support streaming response ([#88](https://github.com/AIGNE-io/aigne-framework/issues/88)) ([4fb4d92](https://github.com/AIGNE-io/aigne-framework/commit/4fb4d92f8b36011437efba3265591b2477f2d680))

## [1.10.0](https://github.com/AIGNE-io/aigne-framework/compare/core-v1.9.0...core-v1.10.0) (2025-04-22)


### Features

* **stream:** add streaming output support for agent ([#73](https://github.com/AIGNE-io/aigne-framework/issues/73)) ([5f3ea4b](https://github.com/AIGNE-io/aigne-framework/commit/5f3ea4bccda7c8c457d6e9518b3d6a8b254ec041))


### Bug Fixes

* **core:** support dynamic model capability detection ([#72](https://github.com/AIGNE-io/aigne-framework/issues/72)) ([9d56d98](https://github.com/AIGNE-io/aigne-framework/commit/9d56d9885778962e5bef806445ad8c4d199f2c65))

## [1.9.0](https://github.com/AIGNE-io/aigne-framework/compare/core-v1.8.0...core-v1.9.0) (2025-04-20)


### Features

* **core:** add model adapters for DeepSeek, Gemini, OpenRouter, and Ollama ([#53](https://github.com/AIGNE-io/aigne-framework/issues/53)) ([5d40546](https://github.com/AIGNE-io/aigne-framework/commit/5d40546bd5ddb70233d27ea3b20e5711b2af320a))


### Bug Fixes

* **dx:** custom error message for agent input/output validation ([#71](https://github.com/AIGNE-io/aigne-framework/issues/71)) ([5145673](https://github.com/AIGNE-io/aigne-framework/commit/5145673aaae2cd6665912e80b1c644e974c42b2f))

## [1.8.0](https://github.com/AIGNE-io/aigne-framework/compare/core-v1.7.0...core-v1.8.0) (2025-04-17)


### Features

* **ci:** support coverage examples with model matrix ([#59](https://github.com/AIGNE-io/aigne-framework/issues/59)) ([1edd704](https://github.com/AIGNE-io/aigne-framework/commit/1edd70426b80a69e3751b2d5fe818297711d0777))
* **cli:** support model and download customization for aigne run ([#61](https://github.com/AIGNE-io/aigne-framework/issues/61)) ([51f6619](https://github.com/AIGNE-io/aigne-framework/commit/51f6619e6c591a84f1f2339b26ef66d89fa9486e))


### Bug Fixes

* **mcp:** set default timeout to 60s ([#67](https://github.com/AIGNE-io/aigne-framework/issues/67)) ([40dc029](https://github.com/AIGNE-io/aigne-framework/commit/40dc029b7795650283a505fd71b9566e5f0a4471))

## [1.7.0](https://github.com/AIGNE-io/aigne-framework/compare/core-v1.6.0...core-v1.7.0) (2025-04-15)


### Features

* add TerminalTracer for better UX in terminal ([#56](https://github.com/AIGNE-io/aigne-framework/issues/56)) ([9875a5d](https://github.com/AIGNE-io/aigne-framework/commit/9875a5d46abb55073340ffae841fed6bd6b83ff4))
* **cli:** support run agents from remote URL ([#60](https://github.com/AIGNE-io/aigne-framework/issues/60)) ([5f49920](https://github.com/AIGNE-io/aigne-framework/commit/5f4992089d36f9e780ba55a912a1d35508cad28e))
* **core:** support oauth for McpAgent with example ([#55](https://github.com/AIGNE-io/aigne-framework/issues/55)) ([9420f3a](https://github.com/AIGNE-io/aigne-framework/commit/9420f3a56cf18986cd45f173044e660be76daab4))


### Bug Fixes

* remove usage of new Node.js exists API for compatibility ([#57](https://github.com/AIGNE-io/aigne-framework/issues/57)) ([c10cc08](https://github.com/AIGNE-io/aigne-framework/commit/c10cc086d8ecd0744f38cdb1367d4c8816b723b3))

## [1.6.0](https://github.com/AIGNE-io/aigne-framework/compare/core-v1.5.0...core-v1.6.0) (2025-04-08)


### Features

* add `serve` command for @aigne/cli ([#54](https://github.com/AIGNE-io/aigne-framework/issues/54)) ([1cca843](https://github.com/AIGNE-io/aigne-framework/commit/1cca843f1760abe832b6651108fa858130f47355))
* add agent library support ([#51](https://github.com/AIGNE-io/aigne-framework/issues/51)) ([1f0d34d](https://github.com/AIGNE-io/aigne-framework/commit/1f0d34ddda3154283a4bc958ddb9b68b4ac106b0))
* support token/call/time limits for ExecutionEngine ([#44](https://github.com/AIGNE-io/aigne-framework/issues/44)) ([5a2ca0a](https://github.com/AIGNE-io/aigne-framework/commit/5a2ca0a033267dd4765f574b53dca71e932e53d4))


### Bug Fixes

* support reconnect to the MCP server automatically ([#50](https://github.com/AIGNE-io/aigne-framework/issues/50)) ([898d83f](https://github.com/AIGNE-io/aigne-framework/commit/898d83f75fc655142b93c70a1afeda376a2e92b4))

## [1.5.0](https://github.com/AIGNE-io/aigne-framework/compare/core-v1.4.0...core-v1.5.0) (2025-03-27)


### Features

* **dx:** show mcp server url on connecting ([#39](https://github.com/AIGNE-io/aigne-framework/issues/39)) ([5819a76](https://github.com/AIGNE-io/aigne-framework/commit/5819a76435fae7333720f9e0c58a25aebc1089e3))


### Bug Fixes

* **dx:** export models/utils in submodules ([#43](https://github.com/AIGNE-io/aigne-framework/issues/43)) ([bd561b3](https://github.com/AIGNE-io/aigne-framework/commit/bd561b397de816f04c2d63d58538e81fba82fc7f))

## [1.4.0](https://github.com/AIGNE-io/aigne-framework/compare/core-v1.3.1...core-v1.4.0) (2025-03-26)


### Features

* **core:** add xAI chat model adapter ([#34](https://github.com/AIGNE-io/aigne-framework/issues/34)) ([b228d22](https://github.com/AIGNE-io/aigne-framework/commit/b228d22b550535ab8e511f13de9e4a65dd73e3c0))


### Bug Fixes

* **orchestrator:** refactor and enhance orchestrator with step synthesis ([#31](https://github.com/AIGNE-io/aigne-framework/issues/31)) ([ba9fca0](https://github.com/AIGNE-io/aigne-framework/commit/ba9fca04fad71d49c8f4f52172b56668a94ea714))

## [1.3.1](https://github.com/AIGNE-io/aigne-framework/compare/core-v1.3.0...core-v1.3.1) (2025-03-25)


### Bug Fixes

* **core:** use system message as user message for claude model if needed ([#32](https://github.com/AIGNE-io/aigne-framework/issues/32)) ([316a6d5](https://github.com/AIGNE-io/aigne-framework/commit/316a6d51f885cceee4020c24695f6588f6b7425f))

## [1.2.0](https://github.com/AIGNE-io/aigne-framework/compare/core-next-v1.1.0...core-next-v1.2.0) (2025-03-18)

- chore: release v1.2.0

## 1.1.0-beta.17 (2025-3-18)

- chore: add support for esm module


## 1.1.0-beta.16 (2025-3-18)

- chore: add puppeteer in linux need docker_container

## 1.1.0-beta.15 (2025-3-18)

- chore: make coverage report as text to terminal
- chore: update contributing docs

## 1.1.0-beta.14 (2025-3-18)

- chore(example): add code-execution example

## 1.1.0-beta.13 (2025-3-18)

- feat: add OrchestratorAgent in agent library

## 1.1.0-beta.12 (2025-3-14)

- chore(example): add concurrency reflection handoff workflow examples

## 1.1.0-beta.11 (2025-3-14)

- feat(core): add sequential and parallel helper function
- chore(examples): add workflow-sequential example

## 1.1.0-beta.10 (2025-3-13)

- chore: ensure required environment variables have values

## 1.1.0-beta.9 (2025-3-13)

- fix(MCP): catch list resource error treat as empty list

## 1.1.0-beta.8 (2025-3-13)

- fix(AIAgent): should pass both arguments (model generated) and input (user input) to tool
- chore(examples): add workflow-router example
- chore(examples): rename examples puppeteer-mcp-server and sqlite-mcp-server to mcp-server-puppeteer and mcp-server-sqlite

## 1.1.0-beta.7 (2025-3-13)

- chore: rename @aigne/core to @aigne/core

## 1.1.0-beta.6 (2025-3-13)

- chore(examples): default enable mcp debug message for examples

## 1.1.0-beta.5 (2025-3-13)

- feat: support chat history in PromptBuilder
- feat: add `prompts` for MCPAgent to consume prompts from MCP server
- chore: add sqlite-mcp-server example
- test: add more unit test cases

## 1.1.0-beta.4 (2025-3-12)

- feat: support run puppeteer example chat loop in terminal

## 1.1.0-beta.3 (2025-3-11)

- chore: set module type for core package

## 1.1.0-beta.2 (2025-3-11)

- feat: use PromptBuilder instead of string instructions
- refactor: use tools instead of skills
- chore(examples): add puppeteer-mcp-server example

## 1.1.0-beta.1 (2025-3-11)

- feat: add Agent FunctionAgent AIAgent MCPAgent and ExecutionEngine
