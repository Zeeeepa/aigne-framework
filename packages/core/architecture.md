## Class Definitions

```mermaid
classDiagram
    class PromptBuilderBuildOptions {
        +Context context
        +Agent agent
        +object input
        +ChatModel model
    }

    class Prompt {
        +List~object~ messages
        +List~Agent~ skills
        +object toolChoice
        +object responseFormat
    }

    PromptBuilder ..> PromptBuilderBuildOptions: dependency
    PromptBuilder ..> Prompt: dependency
    class PromptBuilder {
        +build(PromptBuilderBuildOptions options) Prompt
    }

    ChatModel --|> Agent: inheritance
    class ChatModel {
    }

    ImageModel --|> Agent: inheritance
    class ImageModel {
    }

    class Agent {
        +string name
        +string description
        +object inputSchema
        +object outputSchema
        +List~string~ subscribeTopic
        +List~string~ publishTopic
        +List~Agent~ skills
        +MemoryAgent memory

        +invoke(object input, Context context) object
        +shutdown() void
        +process(object input, Context context) object*
        -preprocess() void
        -postprocess() void
    }

    AIAgent --|> Agent: inheritance
    AIAgent *.. PromptBuilder: composition
    class AIAgent {
        +ChatModel model
        +PromptBuilder instructions
        +string outputKey
        +object toolChoice
        +boolean catchToolsError
    }

    TeamAgent --|> Agent: inheritance
    class TeamAgent {
        +ProcessMode mode
    }

    ImageAgent --|> Agent: inheritance
    ImageAgent *.. PromptBuilder: composition
    ImageAgent *.. ImageModel: composition
    class ImageAgent {
        +ImageModel model
        +PromptBuilder instructions
    }

    FunctionAgent --|> Agent: inheritance
    class FunctionAgent {
        +Function process
    }

    RPCAgent --|> Agent: inheritance
    class RPCAgent {
        +string url
    }

    MCPAgent --|> Agent: inheritance
    MCPAgent *.. MCPClient: composition
    class MCPAgent {
        +MCPClient client
    }

    class MCPClient {
    }

    class Message {
        +object output
    }

    MessageQueue ..> Message: dependency
    class MessageQueue {
        +publish(string topic, Message message) void
        +subscribe(string topic, Function callback) void
        +unsubscribe(string topic, Function callback) void
    }

    Context *.. ChatModel: composition
    Context *.. ImageModel: composition
    Context --|> MessageQueue: inheritance
    class Context {
        +ChatModel model
        +ImageModel imageModel
        +List~Agent~ skills

        +invoke(Agent agent, object input) object
        +publish(string topic, Message message) void
        +subscribe(string topic, Function callback) void
        +unsubscribe(string topic, Function callback) void
    }

    UserAgent --|> Agent: inheritance
    class UserAgent {
    }

    class EventEmitter {
        +on(): void
        +emit(): void
    }

    AIGNE --|> EventEmitter: inheritance
    class AIGNE {
        +string name
        +string description
        +ChatModel model
        +object limits
        +List~Agent~ agents
        +List~Agent~ skills

        +load(string path) AIGNE$
        +addAgent(Agent agent) void
        +invoke(Agent agent) UserAgent
        +invoke(Agent agent, string input) object
        +invoke(Agent agent, object input) object
        +publish(string topic, Message message) void
        +subscribe(string topic, Function callback) void
        +unsubscribe(string topic, Function callback) void
        +shutdown() void
    }
```
