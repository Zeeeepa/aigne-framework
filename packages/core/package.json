{
  "name": "@aigne/core",
  "version": "1.12.0",
  "description": "AIGNE core library for building AI-powered applications",
  "publishConfig": {
    "access": "public"
  },
  "author": "Arcblock <blocklet@arcblock.io> https://github.com/blocklet",
  "homepage": "https://github.com/AIGNE-io/aigne-framework",
  "license": "Elastic-2.0",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/AIGNE-io/aigne-framework"
  },
  "files": [
    "lib/cjs",
    "lib/dts",
    "lib/esm",
    "LICENSE",
    "README.md",
    "CHANGELOG.md"
  ],
  "type": "module",
  "main": "./lib/cjs/index.js",
  "module": "./lib/esm/index.js",
  "types": "./lib/dts/index.d.ts",
  "exports": {
    ".": {
      "import": "./lib/esm/index.js",
      "require": "./lib/cjs/index.js",
      "types": "./lib/dts/index.d.ts"
    },
    "./*": {
      "import": "./lib/esm/*",
      "require": "./lib/cjs/*",
      "types": "./lib/dts/*"
    }
  },
  "typesVersions": {
    "*": {
      ".": [
        "./lib/dts/index.d.ts"
      ],
      "agents/*": [
        "./lib/dts/agents/*"
      ],
      "aigne/*": [
        "./lib/dts/aigne/*"
      ],
      "loader/*": [
        "./lib/dts/loader/*"
      ],
      "models/*": [
        "./lib/dts/models/*"
      ],
      "prompt/*": [
        "./lib/dts/prompt/*"
      ],
      "utils/*": [
        "./lib/dts/utils/*"
      ]
    }
  },
  "scripts": {
    "lint": "tsc --noEmit",
    "build": "tsc --build scripts/tsconfig.build.json",
    "clean": "rimraf lib test/coverage",
    "prepublishOnly": "run-s clean build",
    "build:docs": "typedoc",
    "test": "run-s test:src test:test",
    "test:coverage": "run-s test:src:coverage test:test:coverage",
    "test:test": "bun --cwd test test",
    "test:test:coverage": "bun --cwd test --coverage --coverage-reporter=lcov --coverage-reporter=text",
    "test:src": "bun test src/",
    "test:src:coverage": "bun test src/ --coverage --coverage-reporter=lcov --coverage-reporter=text",
    "postbuild": "echo '{\"type\": \"module\"}' > lib/esm/package.json && echo '{\"type\": \"commonjs\"}' > lib/cjs/package.json"
  },
  "dependencies": {
    "@aigne/json-schema-to-zod": "^1.3.3",
    "@modelcontextprotocol/sdk": "^1.10.2",
    "@types/debug": "^4.1.12",
    "camelize-ts": "^3.0.0",
    "content-type": "^1.0.5",
    "debug": "^4.4.0",
    "eventsource-parser": "^3.0.1",
    "fast-deep-equal": "^3.1.3",
    "immer": "^10.1.1",
    "inquirer": "^12.5.2",
    "mustache": "^4.2.0",
    "nanoid": "^5.1.5",
    "p-retry": "^6.2.1",
    "raw-body": "^3.0.0",
    "ufo": "^1.6.1",
    "uuid": "^11.1.0",
    "yaml": "^2.7.1",
    "zod": "^3.24.2",
    "zod-to-json-schema": "^3.24.5"
  },
  "peerDependencies": {
    "@anthropic-ai/sdk": "^0.39.0",
    "@google/generative-ai": "^0.24.0",
    "openai": "^4.87.3"
  },
  "devDependencies": {
    "@anthropic-ai/sdk": "^0.39.0",
    "@google/generative-ai": "^0.24.0",
    "@microsoft/tsdoc": "^0.15.1",
    "@types/bun": "^1.2.9",
    "@types/compression": "^1.7.5",
    "@types/content-type": "^1.1.8",
    "@types/express": "^5.0.1",
    "@types/mustache": "^4.2.5",
    "@types/node": "^22.14.1",
    "compression": "^1.8.0",
    "detect-port": "^2.1.0",
    "express": "^5.1.0",
    "hono": "^4.7.7",
    "npm-run-all": "^4.1.5",
    "openai": "^4.94.0",
    "prettier": "^3.5.3",
    "rimraf": "^6.0.1",
    "typedoc": "^0.28.3",
    "typedoc-github-wiki-theme": "^2.1.0",
    "typedoc-plugin-markdown": "^4.6.3",
    "typedoc-plugin-no-inherit": "^1.6.1",
    "typescript": "^5.8.3"
  }
}
