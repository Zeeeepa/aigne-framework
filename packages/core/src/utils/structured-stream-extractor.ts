import { type AgentResponseChunk, isAgentResponseDelta } from "../agents/agent.js";
import type { ChatModelOutput } from "../agents/chat-model.js";

export class ExtractMetadataTransform extends TransformStream<
  AgentResponseChunk<ChatModelOutput>,
  AgentResponseChunk<ChatModelOutput & { metadata?: string }>
> {
  private buffer = "";

  private cursor = 0;

  private state: "none" | "start" = "none";

  constructor({
    start,
    end,
    parse,
  }: { start: string; end: string; parse: (raw: string) => object }) {
    super({
      transform: async (chunk, controller) => {
        if (isAgentResponseDelta(chunk) && chunk.delta.text?.text) {
          const text = chunk.delta.text.text;

          this.buffer += text;

          for (;;) {
            if (this.state === "none") {
              const found = findMatchIndex(this.buffer, this.cursor, start);
              if (found.start > this.cursor) {
                let text = this.buffer.slice(this.cursor, found.start);
                this.cursor = found.start;

                // Trim trailing whitespace from the text
                const whitespace = text.match(/(?<whitespace>\s+)$/)?.groups?.whitespace;
                if (whitespace) {
                  text = text.slice(0, -whitespace.length);
                  this.cursor -= whitespace.length;
                }

                if (text) {
                  controller.enqueue({ delta: { text: { text } } });
                }
              }

              if (found.end) {
                this.state = "start";
                this.cursor = found.end;
              }
            }

            if (this.state === "start") {
              const found = findMatchIndex(this.buffer, this.cursor, end);
              if (found.end) {
                const metadata = this.buffer.slice(this.cursor, found.start);
                const json = parse(metadata);
                controller.enqueue({ delta: { json: { json } } });

                this.state = "none";
                this.cursor = found.end;
                continue;
              }
            }

            break;
          }

          return;
        }

        controller.enqueue(chunk);
      },
    });
  }
}

function findMatchIndex(
  str: string,
  position: number,
  match: string,
): { start: number; end?: number } {
  const i = str.indexOf(match, position);
  if (i >= 0) return { start: i, end: i + match.length };

  for (let i = match.length - 1; i > 0; i--) {
    const m = match.slice(0, i);
    if (str.endsWith(m)) {
      return { start: str.length - m.length };
    }
  }

  return { start: str.length };
}
