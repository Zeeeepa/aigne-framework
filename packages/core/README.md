# @aigne/core

[![GitHub star chart](https://img.shields.io/github/stars/AIGNE-io/aigne-framework?style=flat-square)](https://star-history.com/#AIGNE-io/aigne-framework)
[![Open Issues](https://img.shields.io/github/issues-raw/AIGNE-io/aigne-framework?style=flat-square)](https://github.com/AIGNE-io/aigne-framework/issues)
[![codecov](https://codecov.io/gh/AIGNE-io/aigne-framework/graph/badge.svg?token=DO07834RQL)](https://codecov.io/gh/AIGNE-io/aigne-framework)
[![NPM Version](https://img.shields.io/npm/v/@aigne/core)](https://www.npmjs.com/package/@aigne/core)
[![Elastic-2.0 licensed](https://img.shields.io/npm/l/@aigne/core)](https://github.com/AIGNE-io/aigne-framework/blob/main/LICENSE)

**English** | [中文](README.zh.md)

Core library of [AIGNE Framework](https://github.com/AIGNE-io/aigne-framework) for building AI-powered applications.

## Introduction

`@aigne/core` is the foundation component of [AIGNE Framework](https://github.com/AIGNE-io/aigne-framework), providing the essential modules and tools needed to build AI-driven applications. This package implements the core functionalities of the framework, including agent systems, aigne environment, model integrations, and workflow pattern support.

## Features

- **Multiple AI Model Support**: Built-in support for OpenAI, Gemini, Claude, Nova, and other mainstream AI models, easily extensible to support additional models
- **Agent System**: Powerful agent abstractions supporting AI agents, function agents, MCP agents, and more
- **AIGNE Environment**: Flexible handling communication between agents and workflow execution
- **Workflow Patterns**: Support for sequential, concurrent, routing, handoff, and other workflow patterns
- **MCP Protocol Integration**: Seamless integration with external systems through the Model Context Protocol
- **TypeScript Support**: Comprehensive type definitions providing an excellent development experience

## Installation

```bash
# Using npm
npm install @aigne/core

# Using yarn
yarn add @aigne/core

# Using pnpm
pnpm add @aigne/core
```

## Basic Usage

```typescript
import { AIAgent, AIGNE } from "@aigne/core";
import { OpenAIChatModel } from "@aigne/core/models/openai-chat-model.js";

// Create AI model instance
const model = new OpenAIChatModel({
  apiKey: process.env.OPENAI_API_KEY,
  model: process.env.DEFAULT_CHAT_MODEL || "gpt-4-turbo",
});

// Create AI agent
const agent = AIAgent.from({
  name: "Assistant",
  instructions: "You are a helpful assistant.",
});

// Create AIGNE
const aigne = new AIGNE({ model });

// Use the AIGNE to invoke the agent
const userAgent = await aigne.invoke(agent);

// Send a message to the agent
const response = await userAgent.invoke("Hello, can you help me write a short article?");
console.log(response);
```

## Module Structure

- `agents/`: Agent implementations, including AI agents, function agents, MCP agents, etc.
- `aigne/`: AIGNE is agents execution environment
- `loader/`: Loader-related functionality
- `models/`: Integration with various AI models
- `prompt/`: Prompt handling functionality
- `utils/`: Utility functions and helper methods

## Documentation

For more detailed API documentation, please refer to:

- [Agent API](../../docs/apis/agent-api.md)
- [AI Agent API](../../docs/apis/ai-agent-api.md)
- [Function Agent API](../../docs/apis/function-agent-api.md)
- [MCP Agent API](../../docs/apis/mcp-agent-api.md)
- [AIGNE API](../../docs/apis/aigne-api.md)
- [Server/Client API](../../docs/apis/server-client-api.md)

## License

Elastic-2.0
