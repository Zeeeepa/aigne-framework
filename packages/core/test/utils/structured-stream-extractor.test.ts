import { expect, test } from "bun:test";
import { readableStreamToArray, stringToAgentResponseStream } from "@aigne/core";
import { ExtractMetadataTransform } from "@aigne/core/utils/structured-stream-extractor.js";

test("ExtractMetadataTransform should extract metadata from structured stream", async () => {
  const stream = stringToAgentResponseStream(`\
Hello, How can I help you today?<metadata>
{
  "foo": "bar",
  "baz": 123
}
</metadata>`).pipeThrough(
    new ExtractMetadataTransform({
      start: "<metadata>",
      end: "</metadata>",
      parse: JSON.parse,
    }),
  );

  expect(readableStreamToArray(stream)).resolves.toMatchSnapshot();
});

test("ExtractMetadataTransform should trim trailing whitespace", async () => {
  const stream = stringToAgentResponseStream(`\
Hello, How can I help you today?

<metadata>
{
  "foo": "bar",
  "baz": 123
}
</metadata>`).pipeThrough(
    new ExtractMetadataTransform({
      start: "<metadata>",
      end: "</metadata>",
      parse: JSON.parse,
    }),
  );

  expect(readableStreamToArray(stream)).resolves.toMatchSnapshot();
});
