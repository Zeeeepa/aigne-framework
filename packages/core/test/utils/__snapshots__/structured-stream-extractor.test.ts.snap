// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`ExtractMetadataTransform should extract metadata from structured stream 1`] = `
[
  {
    "delta": {
      "text": {
        "text": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " How",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " can",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " I",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " help",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " you",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " today",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "?",
      },
    },
  },
  {
    "delta": {
      "json": {
        "json": {
          "baz": 123,
          "foo": "bar",
        },
      },
    },
  },
]
`;

exports[`ExtractMetadataTransform should trim trailing whitespace 1`] = `
[
  {
    "delta": {
      "text": {
        "text": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " How",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " can",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " I",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " help",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " you",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " today",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "?",
      },
    },
  },
  {
    "delta": {
      "json": {
        "json": {
          "baz": 123,
          "foo": "bar",
        },
      },
    },
  },
]
`;
